lib/lists.nix

```diff
@@ -403,6 +403,45 @@ rec {
   */
   concatMap = builtins.concatMap;
 
+  /**
+    Map, removing null values.
+
+    # Type
+
+    ```
+    filterMap :: (a -> [b]) -> [a] -> [b]
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.lists.filterMap` usage example
+
+    ```nix
+    let f = x:
+      if "udp" == x.proto
+      then x.port
+      else null;
+    in filterMap f [
+      {
+        dns.proto  = "udp";
+        dns.port  = 53;
+      }
+      {
+        http.proto = "tcp";
+        http.port = 80;
+      }
+    ]
+    => [ 53 ]
+    ```
+
+    :::
+  */
+  filterMap = f:
+    let g = x: acc:
+      let y = f x;
+      in if isNull y then acc else [y] ++ acc;
+    in lib.foldr g [];
```

Comment:
We're going to be strict in the whole thing anyway, realistically, so we might as well `foldl'` to be quicker and constant-stack.
At that point, it can be compared to a solution based on `concatMap`. That's a builtin, so it should be pretty fast.

Reply:
are you refering to `lib.foldl'` or `builtins.foldl'`. I assume the former?

Reply:
Good question. I think in this case we don't need the `lib.foldl'` improvement, because we don't need to "evaluating `[]` first" does nothing.

Normally it is a good idea to use `lib` instead of `builtins`, but in this case we don't need it, and we can squeeze a tiny bit of extra performance. It really only makes sense for functions in a heavily used and/or performance critical library like `lib`  to even consider to use the `builtins` function instead of the usually better `lib`.